File: app/src/main/java/de/christinecoenen/code/zapp/app/MainActivity.java
package de.christinecoenen.code.zapp.app;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.preference.PreferenceManager;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.bottomnavigation.BottomNavigationView;

import de.christinecoenen.code.zapp.R;
import de.christinecoenen.code.zapp.app.about.ui.AboutActivity;
import de.christinecoenen.code.zapp.app.livestream.ui.list.ChannelListFragment;
import de.christinecoenen.code.zapp.app.mediathek.repository.MediathekSearchSuggestionsProvider;
import de.christinecoenen.code.zapp.app.mediathek.ui.list.MediathekListFragment;
import de.christinecoenen.code.zapp.app.settings.ui.SettingsActivity;
import de.christinecoenen.code.zapp.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity implements ViewPager.OnPageChangeListener, SearchView.OnQueryTextListener {

	private static final String ARG_QUERY = "ARG_QUERY";
	private static final int PAGE_CHANNEL_LIST = 0;
	private static final int PAGE_MEDIATHEK_LIST = 1;

	private ViewPager viewPager;
	private SearchView searchView;
	private BottomNavigationView bottomNavigation;

	private String searchQuery;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		PreferenceManager.setDefaultValues(this, R.xml.preferences, false);

		ActivityMainBinding binding = ActivityMainBinding.inflate(getLayoutInflater());
		setContentView(binding.getRoot());

		viewPager = binding.viewPager;
		searchView = binding.search;
		bottomNavigation = binding.bottomNavigation;

		setSupportActionBar(binding.toolbar);

		viewPager.setAdapter(new MainPageAdapter(getSupportFragmentManager()));
		viewPager.addOnPageChangeListener(this);

		searchView.setOnQueryTextListener(this);
		SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
		searchView.setIconified(false);
		searchView.setIconifiedByDefault(false);
		searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
		searchView.clearFocus();
		searchView.setOnQueryTextFocusChangeListener(this::onSearchQueryTextFocusChangeListener);

		binding.bottomNavigation.setOnNavigationItemSelectedListener(this::onNavigationItemSelected);

		onPageSelected(viewPager.getCurrentItem());

		handleIntent(getIntent());
	}

	@Override
	protected void onNewIntent(Intent intent) {
		super.onNewIntent(intent);
		handleIntent(intent);
	}

	@Override
	protected void onSaveInstanceState(@NonNull Bundle outState) {
		super.onSaveInstanceState(outState);
		outState.putString(ARG_QUERY, searchQuery);
	}

	@Override
	protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
		super.onRestoreInstanceState(savedInstanceState);
		searchQuery = savedInstanceState.getString(ARG_QUERY);
		search(searchQuery);
	}

	@Override
	public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
	}

	@Override
	public void onPageSelected(int position) {
		searchView.clearFocus();

		switch (position) {
			case PAGE_MEDIATHEK_LIST:
				searchView.setVisibility(View.VISIBLE);
				bottomNavigation.setSelectedItemId(R.id.menu_mediathek);
				break;
			case PAGE_CHANNEL_LIST:
			default:
				searchView.setVisibility(View.GONE);
				bottomNavigation.setSelectedItemId(R.id.menu_live);
				break;
		}
	}

	@Override
	public void onPageScrollStateChanged(int state) {
	}

	private void handleIntent(Intent intent) {
		if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
			// called by searchView on search commit

			String query = intent.getStringExtra(SearchManager.QUERY);

			searchView.setOnQueryTextListener(null);
			searchView.clearFocus();
			searchView.setQuery(query, false);
			searchView.setOnQueryTextListener(this);

			search(query);

			MediathekSearchSuggestionsProvider.saveQuery(this, query);
		}
	}

	private void search(String query) {
		searchQuery = query;

		Fragment currentFragment = getSupportFragmentManager()
			.findFragmentByTag("android:switcher:" + R.id.view_pager + ":" + viewPager.getCurrentItem());

		if (currentFragment instanceof MediathekListFragment) {
			((MediathekListFragment) currentFragment).search(query);
		}
	}

	private boolean onNavigationItemSelected(MenuItem menuItem) {
		Intent intent;

		switch (menuItem.getItemId()) {
			case R.id.menu_live:
				viewPager.setCurrentItem(PAGE_CHANNEL_LIST, false);
				return true;
			case R.id.menu_mediathek:
				viewPager.setCurrentItem(PAGE_MEDIATHEK_LIST, false);
				return true;
			case R.id.menu_about:
				intent = AboutActivity.getStartIntent(this);
				startActivity(intent);
				return true;
			case R.id.menu_settings:
				intent = SettingsActivity.getStartIntent(this);
				startActivity(intent);
				return true;
		}

		return false;
	}

	/*
	 * To open up the soft keyboard on Android (Fire) TV when focusing the SearchView.
	 */
	private void onSearchQueryTextFocusChangeListener(View searchView, boolean hasFocus) {
		if (hasFocus && !searchView.isInTouchMode()) {
			searchView.post(() -> {
				InputMethodManager imm = (InputMethodManager) MainActivity.this.getSystemService(Context.INPUT_METHOD_SERVICE);
				imm.showSoftInput(searchView.findFocus(), InputMethodManager.SHOW_FORCED);
			});
		}
	}

	@Override
	public boolean onQueryTextSubmit(String query) {
		return false;
	}

	@Override
	public boolean onQueryTextChange(String newText) {
		search(newText);
		return true;
	}

	private static class MainPageAdapter extends FragmentPagerAdapter {

		MainPageAdapter(FragmentManager fragmentManager) {
			super(fragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
		}

		@Override
		public int getCount() {
			return 2;
		}

		@NonNull
		@Override
		public Fragment getItem(int position) {
			switch (position) {
				case PAGE_CHANNEL_LIST:
					return ChannelListFragment.getInstance();
				case PAGE_MEDIATHEK_LIST:
				default:
					return MediathekListFragment.getInstance();
			}
		}
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/app/main/MainActivity.kt
package de.christinecoenen.code.zapp.app.main

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.speech.RecognizerIntent
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.inputmethod.InputMethodManager
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.os.bundleOf
import androidx.core.view.MenuProvider
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.isVisible
import androidx.core.view.updateLayoutParams
import androidx.core.widget.addTextChangedListener
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.navigation.NavController
import androidx.navigation.NavDestination
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.navOptions
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI
import androidx.navigation.ui.NavigationUiSaveStateControl
import androidx.navigation.ui.navigateUp
import androidx.navigation.ui.onNavDestinationSelected
import androidx.navigation.ui.setupActionBarWithNavController
import androidx.preference.PreferenceManager
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.search.SearchView
import de.christinecoenen.code.zapp.R
import de.christinecoenen.code.zapp.app.search.SearchViewModel
import de.christinecoenen.code.zapp.app.settings.repository.SettingsRepository
import de.christinecoenen.code.zapp.databinding.ActivityMainBinding
import de.christinecoenen.code.zapp.utils.system.LifecycleOwnerHelper.launchOnCreated
import de.christinecoenen.code.zapp.utils.system.SystemUiHelper.applyHorizontalInsetsAsPadding
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.drop
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel


class MainActivity : AppCompatActivity(), MenuProvider {

	private var _binding: ActivityMainBinding? = null
	private val binding get() = _binding!!

	private val settingsRepository: SettingsRepository by inject()

	private val searchViewModel: SearchViewModel by viewModel()

	private lateinit var navController: NavController
	private lateinit var appBarConfiguration: AppBarConfiguration

	private val requestPermissionLauncher =
		registerForActivityResult(ActivityResultContracts.RequestPermission()) {}

	private val voiceInputLauncher =
		registerForActivityResult(
			ActivityResultContracts.StartActivityForResult(),
			::onVoiceInputReceived
		)

	private val searchViewTransistionListener = SearchView.TransitionListener { _, _, newState ->
		val isHidden = newState == SearchView.TransitionState.HIDDEN
		val isShown = newState == SearchView.TransitionState.SHOWN

		updateBottomNavigationVisibility()

		// set correct state when user shows the search view by clicking the search bar
		if (isShown) {
			searchViewModel.enterLastSearch()
		}

		// set correct state when user pressed back button on searchView
		if (isHidden) {
			when (navController.currentDestination?.id) {
				R.id.searchResultsFragment -> {}
				else -> searchViewModel.exitToNone()
			}
		}
	}

	@OptIn(NavigationUiSaveStateControl::class)
	override fun onCreate(savedInstanceState: Bundle?) {
		enableEdgeToEdge()

		super.onCreate(savedInstanceState)

		_binding = ActivityMainBinding.inflate(layoutInflater)

		setContentView(binding.root)

		val startFragmentId = settingsRepository.startFragment

		navController = binding.navHostFragment.getFragment<NavHostFragment>().navController
		navController.graph.setStartDestination(startFragmentId)

		appBarConfiguration = AppBarConfiguration(
			setOf(
				R.id.channelListFragment,
				R.id.mediathekListFragment,
				R.id.personalFragment,
			),
			fallbackOnNavigateUpListener = ::onSupportNavigateUp
		)

		setSupportActionBar(binding.toolbar)
		setupActionBarWithNavController(navController, appBarConfiguration)

		navController.addOnDestinationChangedListener(::onDestinationChanged)

		NavigationUI.setupWithNavController(binding.bottomNavigation, navController, false)

		binding.searchView.let {
			it.addTransitionListener(searchViewTransistionListener)
			it.editText.addTextChangedListener { editable ->
				if (searchViewModel.searchState.value == SearchViewModel.SeachState.Query) {
					searchViewModel.setSearchQuery(editable.toString())
				}
			}
			it.editText.setOnEditorActionListener { _, _, _ ->
				onSubmitSearch()
				return@setOnEditorActionListener false
			}
			it.setOnMenuItemClickListener {
				startVoiceSearch()
				true
			}
			it.inflateMenu(R.menu.search)
		}

		launchOnCreated {
			searchViewModel.searchQuery.collectLatest {
				if (binding.searchView.text.toString() != it) {
					setSearchViewQueryAndFocus(it)
				}
			}
		}

		launchOnCreated {
			searchViewModel.searchState.drop(1).collectLatest { searchState ->
				val query = binding.searchView.text.toString()

				when (searchState) {
					SearchViewModel.SeachState.None -> {
						binding.searchView.hide()
						binding.searchbar.setText("")
					}

					SearchViewModel.SeachState.Query -> {
						binding.searchView.show()
						setSearchViewQueryAndFocus(query)
					}

					SearchViewModel.SeachState.Results -> {
						binding.searchView.hide()
						binding.searchbar.setText(query)

						navController.navigate(R.id.searchResultsFragment,
							bundleOf("title" to query),
							navOptions { launchSingleTop = true }
						)
					}
				}
			}
		}

		addMenuProvider(this)
		binding.searchbar.addMenuProvider(this)

		PreferenceManager.setDefaultValues(application, R.xml.preferences, false)

		if (savedInstanceState == null &&
			navController.currentDestination?.id == R.id.channelListFragment
		) {
			// new app start - explicitly navigate to start fragment defined in settings
			navController.navigate(startFragmentId)
		}

		requestPermissions()

		binding.appBar.applyHorizontalInsetsAsPadding()
		binding.navHostFragment.applyHorizontalInsetsAsPadding()

		ViewCompat.setOnApplyWindowInsetsListener(binding.toolbar) { v, insets ->
			val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
			v.updateLayoutParams<AppBarLayout.LayoutParams> {
				topMargin = systemBars.top
			}
			insets
		}
	}

	private fun setSearchViewQueryAndFocus(query: String) {
		binding.searchView.setText(query)

		binding.searchView.editText.let {
			it.requestFocus()
			it.setSelection(query.length)

			// re-show onscreen keyboard
			val inputMethodManager = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
			inputMethodManager.showSoftInput(it, InputMethodManager.SHOW_IMPLICIT)
		}
	}

	@Suppress("UNUSED_PARAMETER")
	private fun onDestinationChanged(
		controller: NavController,
		destination: NavDestination,
		arguments: Bundle?
	) {
		// show bottom navigation for main destinations
		updateBottomNavigationVisibility()

		// show search for non destinations
		val showSearchBar = arguments?.getBoolean("show_search_bar", false) == true
		binding.searchbar.isVisible = showSearchBar
	}

	override fun onDestroy() {
		super.onDestroy()
		navController.removeOnDestinationChangedListener(::onDestinationChanged)
		_binding?.searchView?.removeTransitionListener(searchViewTransistionListener)
		_binding = null
	}

	fun addMenuProviderToSearchBar(
		provider: MenuProvider,
		owner: LifecycleOwner,
		state: Lifecycle.State
	) {
		binding.searchbar.addMenuProvider(provider, owner, state)
	}

	override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
		// done by child fragments
	}

	override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
		return menuItem.onNavDestinationSelected(navController)
	}

	override fun onSupportNavigateUp(): Boolean {
		return navController.navigateUp(appBarConfiguration) || super.onSupportNavigateUp()
	}

	private fun onSubmitSearch() {
		binding.fragmentSearch.requestFocus()
		searchViewModel.submit()
	}

	private fun startVoiceSearch() {
		val voiceInputIntent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
			putExtra(
				RecognizerIntent.EXTRA_LANGUAGE_MODEL,
				RecognizerIntent.LANGUAGE_MODEL_FREE_FORM
			)
		}
		voiceInputLauncher.launch(voiceInputIntent)
	}

	private fun onVoiceInputReceived(result: ActivityResult) {
		if (result.resultCode == RESULT_OK) {
			val spokenText = result.data?.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)
				.let { it?.get(0) ?: "" }
			searchViewModel.setSearchQuery(spokenText)
			searchViewModel.submit()
		}
	}

	private fun requestPermissions() {
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
			if (ContextCompat.checkSelfPermission(
					this,
					Manifest.permission.POST_NOTIFICATIONS
				) == PackageManager.PERMISSION_GRANTED
			) {
				return
			}

			requestPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
		}
	}

	private fun updateBottomNavigationVisibility() {
		val arguments = navController.currentDestination?.arguments
		val isMainDestination = arguments?.get("show_bottom_navigation")?.defaultValue == true

		val isSearchOverlayVisible =
			binding.searchView.currentTransitionState == SearchView.TransitionState.SHOWN ||
				binding.searchView.currentTransitionState == SearchView.TransitionState.SHOWING

		binding.bottomNavigation.isVisible = isMainDestination && !isSearchOverlayVisible
	}
}
